File: java/yass/main/ch/softappeal/yass/transport/MessageSerializer.java
package ch.softappeal.yass.transport;

import ch.softappeal.yass.core.remote.ExceptionReply;
import ch.softappeal.yass.core.remote.Message;
import ch.softappeal.yass.core.remote.Request;
import ch.softappeal.yass.core.remote.ValueReply;
import ch.softappeal.yass.serialize.CompositeSerializer;
import ch.softappeal.yass.serialize.Reader;
import ch.softappeal.yass.serialize.Serializer;
import ch.softappeal.yass.serialize.Writer;

import java.util.List;

/**
 * Default {@link Serializer} for {@link Message}.
 */
public final class MessageSerializer extends CompositeSerializer {

    public MessageSerializer(final Serializer contractSerializer) {
        super(contractSerializer);
    }

    private static final byte REQUEST = (byte)0;
    private static final byte VALUE_REPLY = (byte)1;
    private static final byte EXCEPTION_REPLY = (byte)2;

    @SuppressWarnings("unchecked")
    @Override public Message read(final Reader reader) throws Exception {
        final var type = reader.readByte();
        if (type == REQUEST) {
            return new Request(
                reader.readZigZagInt(),
                reader.readZigZagInt(),
                (List<Object>)serializer.read(reader)
            );
        } else if (type == VALUE_REPLY) {
            return new ValueReply(
                serializer.read(reader)
            );
        } else {
            return new ExceptionReply(
                (Exception)serializer.read(reader)
            );
        }
    }

    @Override public void write(final Object message, final Writer writer) throws Exception {
        if (message instanceof Request) {
            writer.writeByte(REQUEST);
            final var request = (Request)message;
            writer.writeZigZagInt(request.serviceId);
            writer.writeZigZagInt(request.methodId);
            serializer.write(request.arguments, writer);
        } else if (message instanceof ValueReply) {
            writer.writeByte(VALUE_REPLY);
            final var reply = (ValueReply)message;
            serializer.write(reply.value, writer);
        } else {
            writer.writeByte(EXCEPTION_REPLY);
            final var reply = (ExceptionReply)message;
            serializer.write(reply.exception, writer);
        }
    }

}

File: kotlin/yass/main/ch/softappeal/yass/transport/MessageSerializer.kt
package ch.softappeal.yass.transport

import ch.softappeal.yass.remote.*
import ch.softappeal.yass.serialize.*

private const val Request = 0.toByte()
private const val ValueReply = 1.toByte()
private const val ExceptionReply = 2.toByte()

fun messageSerializer(contractSerializer: Serializer) = object : Serializer {
    override fun read(reader: Reader): Message = when (val type = reader.readByte()) {
        Request -> Request(
            reader.readZigZagInt(),
            reader.readZigZagInt(),
            contractSerializer.read(reader) as List<Any?>
        )
        ValueReply -> ValueReply(
            contractSerializer.read(reader)
        )
        ExceptionReply -> ExceptionReply(
            contractSerializer.read(reader) as Exception
        )
        else -> error("unexpected type $type")
    }

    override fun write(writer: Writer, value: Any?) = when (value) {
        is Request -> {
            writer.writeByte(Request)
            writer.writeZigZagInt(value.serviceId)
            writer.writeZigZagInt(value.methodId)
            contractSerializer.write(writer, value.arguments)
        }
        is ValueReply -> {
            writer.writeByte(ValueReply)
            contractSerializer.write(writer, value.value)
        }
        is ExceptionReply -> {
            writer.writeByte(ExceptionReply)
            contractSerializer.write(writer, value.exception)
        }
        else -> error("unexpected value '$value'")
    }
}
